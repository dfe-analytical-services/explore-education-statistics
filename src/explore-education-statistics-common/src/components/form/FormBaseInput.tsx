import ErrorMessage from '@common/components/ErrorMessage';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  memo,
  MouseEventHandler,
  ReactNode,
} from 'react';

export interface FormBaseInputProps extends FormLabelProps {
  addOn?: ReactNode;
  addOnContainerClassName?: string;
  className?: string;
  disabled?: boolean;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  list?: string;
  name: string;
  width?: 20 | 10 | 5 | 4 | 3 | 2;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onClick?: MouseEventHandler<HTMLInputElement>;
  onKeyPress?: KeyboardEventHandler<HTMLInputElement>;
}

interface HiddenProps {
  defaultValue?: string | number;
  type?: 'text' | 'number' | 'color' | 'search';
  value?: string | number;
}

const FormBaseInput = ({
  addOn,
  addOnContainerClassName,
  className,
  error,
  hint,
  id,
  hideLabel,
  label,
  labelSize,
  width,
  type = 'text',
  ...props
}: FormBaseInputProps & HiddenProps) => {
  const input = (
    <input
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      aria-describedby={
        classNames({
          [`${id}-error`]: !!error,
          [`${id}-hint`]: !!hint,
        }) || undefined
      }
      className={classNames('govuk-input', className, {
        [`govuk-input--width-${width}`]: width !== undefined,
        'govuk-input--error': !!error,
      })}
      id={id}
      type={type}
    />
  );

  return (
    <>
      <FormLabel
        id={id}
        label={label}
        labelSize={labelSize}
        hideLabel={hideLabel}
      />

      {hint && (
        <span id={`${id}-hint`} className="govuk-hint">
          {hint}
        </span>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
      {addOn ? (
        <div className={classNames('dfe-flex', addOnContainerClassName)}>
          {input}
          {addOn}
        </div>
      ) : (
        input
      )}
    </>
  );
};

export default memo(FormBaseInput);
