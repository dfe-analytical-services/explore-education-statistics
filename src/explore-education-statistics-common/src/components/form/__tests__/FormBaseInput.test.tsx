import { render } from '@testing-library/react';
import React from 'react';
import FormBaseInput from '../FormBaseInput';

describe('FormTextInput', () => {
  test('renders correctly with required props', () => {
    const { container, getByLabelText } = render(
      <FormBaseInput id="test-input" label="Test input" name="testInput" />,
    );

    expect(getByLabelText('Test input')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with hint', () => {
    const { container, getByText } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
      />,
    );

    const hint = getByText('Fill me in');

    expect(hint.id).toBe('test-input-hint');
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with error', () => {
    const { container, getByText } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        error="Field is required"
      />,
    );

    const error = getByText('Field is required');

    expect(error.id).toBe('test-input-error');
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('aria-describedby is equal to the hint id', () => {
    const { getByLabelText, getByText } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-hint',
    );
  });

  test('aria-describedby is equal to the error id', () => {
    const { getByLabelText, getByText } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        error="Field is required"
      />,
    );

    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-error',
    );
  });

  test('aria-describedby contains both hint and error ids', () => {
    const { getByLabelText, getByText } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        error="Field is required"
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );

    const ariaDescribedBy =
      getByLabelText('Test input').getAttribute('aria-describedby');

    expect(ariaDescribedBy).toContain('test-input-error');
    expect(ariaDescribedBy).toContain('test-input-hint');
  });

  test('renders with a specific width class', () => {
    const { container } = render(
      <FormBaseInput
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        width={20}
      />,
    );

    expect(container.querySelector('.govuk-input--width-20')).not.toBeNull();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders with an add on', () => {
    const { container, getByRole } = render(
      <FormBaseInput
        addOn={<button type="button">Click me!</button>}
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        width={20}
      />,
    );

    expect(getByRole('button', { name: 'Click me!' })).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });
});
