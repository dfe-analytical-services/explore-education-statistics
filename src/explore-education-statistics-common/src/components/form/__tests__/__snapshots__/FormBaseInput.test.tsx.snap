// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextInput renders correctly with error 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span class="govuk-error-message"
      id="test-input-error"
>
  <span class="govuk-visually-hidden">
    Error:
  </span>
  Field is required
</span>
<input name="testInput"
       aria-describedby="test-input-error"
       class="govuk-input govuk-input--error"
       id="test-input"
       type="text"
>
`;

exports[`FormTextInput renders correctly with hint 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span id="test-input-hint"
      class="govuk-hint"
>
  Fill me in
</span>
<input name="testInput"
       aria-describedby="test-input-hint"
       class="govuk-input"
       id="test-input"
       type="text"
>
`;

exports[`FormTextInput renders correctly with required props 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<input name="testInput"
       class="govuk-input"
       id="test-input"
       type="text"
>
`;

exports[`FormTextInput renders with a specific width class 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span id="test-input-hint"
      class="govuk-hint"
>
  Fill me in
</span>
<input name="testInput"
       aria-describedby="test-input-hint"
       class="govuk-input govuk-input--width-20"
       id="test-input"
       type="text"
>
`;

exports[`FormTextInput renders with an add on 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span id="test-input-hint"
      class="govuk-hint"
>
  Fill me in
</span>
<div class="dfe-flex">
  <input name="testInput"
         aria-describedby="test-input-hint"
         class="govuk-input govuk-input--width-20"
         id="test-input"
         type="text"
  >
  <button type="button">
    Click me!
  </button>
</div>
`;
